File: frontend/src/services/analytics.ts
import { api } from '@/services/api'
import type {
  AnalyticsBreakdownsResponse,
  AnalyticsFilters,
  AnalyticsStatisticsResponse,
  AnalyticsTradesResponse,
} from '@/types/analytics'

/**
 * Turn an `AnalyticsFilters` object into `URLSearchParams` entries.
 *
 * Array filters (asset_ids, tag_ids, emotion_ids, entry_timeframe, types) are
 * joined as comma-separated strings — the backend's `filter_params` dependency
 * expects `Optional[str]` and splits on commas.
 * Undefined / empty values are omitted entirely.
 */
function filtersToParams(filters: AnalyticsFilters): URLSearchParams {
  const params = new URLSearchParams()

  if (filters.date_from) params.set('date_from', filters.date_from)
  if (filters.date_to) params.set('date_to', filters.date_to)
  if (filters.asset_ids && filters.asset_ids.length > 0) {
    params.set('asset_ids', filters.asset_ids.join(','))
  }
  if (filters.direction) params.set('direction', filters.direction)
  if (filters.entry_timeframe && filters.entry_timeframe.length > 0) {
    params.set('entry_timeframe', filters.entry_timeframe.join(','))
  }
  if (filters.tag_ids && filters.tag_ids.length > 0) {
    params.set('tag_ids', filters.tag_ids.join(','))
  }
  if (filters.tags_logic) params.set('tags_logic', filters.tags_logic)
  if (filters.emotion_ids && filters.emotion_ids.length > 0) {
    params.set('emotion_ids', filters.emotion_ids.join(','))
  }
  if (filters.types && filters.types.length > 0) {
    params.set('types', filters.types.join(','))
  }
  if (filters.include_missed !== undefined) {
    params.set('include_missed', String(filters.include_missed))
  }
  if (filters.pnl_operator) params.set('pnl_operator', filters.pnl_operator)
  if (filters.pnl_value !== undefined) params.set('pnl_value', String(filters.pnl_value))
  if (filters.duration_operator) params.set('duration_operator', filters.duration_operator)
  if (filters.duration_value !== undefined) {
    params.set('duration_value', String(filters.duration_value))
  }
  if (filters.duration_unit) params.set('duration_unit', filters.duration_unit)

  return params
}

/** Fetch aggregate KPI statistics for the given filters. */
export function fetchStatistics(
  filters: AnalyticsFilters,
  signal?: AbortSignal,
): Promise<AnalyticsStatisticsResponse> {
  const params = filtersToParams(filters)
  const qs = params.toString()
  return api.get<AnalyticsStatisticsResponse>(
    `/analytics/statistics${qs ? `?${qs}` : ''}`,
    signal,
  )
}

/** Fetch the paginated trade list for the given filters. */
export function fetchTrades(
  filters: AnalyticsFilters,
  page: number,
  perPage: number,
  signal?: AbortSignal,
): Promise<AnalyticsTradesResponse> {
  const params = filtersToParams(filters)
  params.set('page', String(page))
  params.set('per_page', String(perPage))
  return api.get<AnalyticsTradesResponse>(`/analytics/trades?${params.toString()}`, signal)
}

/** Fetch chart breakdown data for the given filters. */
export function fetchBreakdowns(
  filters: AnalyticsFilters,
  signal?: AbortSignal,
): Promise<AnalyticsBreakdownsResponse> {
  const params = filtersToParams(filters)
  const qs = params.toString()
  return api.get<AnalyticsBreakdownsResponse>(
    `/analytics/breakdowns${qs ? `?${qs}` : ''}`,
    signal,
  )
}

File: backend/app/routers/analytics.py
from typing import Any, Optional

from fastapi import APIRouter, Depends, Query

from app.models.analytics import (
  AnalyticsStatistics,
  AnalyticsStatisticsResponse,
  AnalyticsTrades,
  AnalyticsTradesResponse,
  DurationUnit,
  RangeOperator,
  SortField,
  SortOrder,
  TagsLogic,
)
from app.services import analytics_service

router = APIRouter(prefix="/api/analytics", tags=["analytics"])


def filter_params(
  date_from: Optional[str] = None,
  date_to: Optional[str] = None,
  asset_ids: Optional[str] = None,
  direction: Optional[str] = None,
  entry_timeframe: Optional[str] = None,
  tag_ids: Optional[str] = None,
  tags_logic: TagsLogic = TagsLogic.and_,
  emotion_ids: Optional[str] = None,
  types: Optional[str] = None,
  include_missed: bool = False,
  pnl_operator: Optional[RangeOperator] = None,
  pnl_value: Optional[float] = None,
  duration_operator: Optional[RangeOperator] = None,
  duration_value: Optional[int] = None,
  duration_unit: DurationUnit = DurationUnit.minutes,
) -> dict[str, Any]:
  """Collect the shared filter query parameters into a single dict."""
  return {
    "date_from": date_from,
    "date_to": date_to,
    "asset_ids": asset_ids,
    "direction": direction,
    "entry_timeframe": entry_timeframe,
    "tag_ids": tag_ids,
    "tags_logic": tags_logic,
    "emotion_ids": emotion_ids,
    "types": types,
    "include_missed": include_missed,
    "pnl_operator": pnl_operator,
    "pnl_value": pnl_value,
    "duration_operator": duration_operator,
    "duration_value": duration_value,
    "duration_unit": duration_unit,
  }


@router.get("/statistics")
async def get_statistics(
  params: dict[str, Any] = Depends(filter_params),
) -> AnalyticsStatisticsResponse:
  data = await analytics_service.get_statistics(params)
  return AnalyticsStatisticsResponse(data=AnalyticsStatistics(**data))


@router.get("/trades")
async def get_trades(
  params: dict[str, Any] = Depends(filter_params),
  page: int = Query(1, ge=1),
  per_page: int = Query(20, ge=1, le=200),
  sort_by: SortField = SortField.trade_date,
  sort_order: SortOrder = SortOrder.desc,
) -> AnalyticsTradesResponse:
  data = await analytics_service.get_trades(
    params, page, per_page, sort_by.value, sort_order.value
  )
  return AnalyticsTradesResponse(data=AnalyticsTrades(**data))
